### PR TITLE
add docker image for cruise-control

### PR DESCRIPTION
**Pull request by zmalik**
This PR adds a docker image for cruise-control.
To enable multiple docker images in the repository. The images/build.sh has been refactored
to support multiple docker images. As for now for kafka and cruise-control

`build.sh --image=kafka` to build kafka image
`build.sh --image=cruise-control` to build cruise-control image
`build.sh --image=kafka --push` to build and push an image

Once the PR is ready to merge, corresponding teamcity changes will be done for the docker build for Kafka and create a new build for Cruise control image. 

Signed-off-by: Zain Malik <zmalikshxil@gmail.com>

### CODE REVIEW

**Comment by mpereira** [comment]
Does it make sense to build all images by default? And push all images when the "push" flag is given? That's how it's being done in [KUDO C*](https://github.com/mesosphere/kudo-cassandra-operator/blob/master/images/build.sh), fwiw.

**Comment by zmalik** [comment]
@mpereira I'm expecting one container image to build/pushed quite less compared to the main base tech container image. So that's why I was going for having two separate builds WDYT?

**Comment by mpereira** [comment]
I think it's probably fine either way. What I was thinking is, given Docker image layer build and push caching I wouldn't worry too much about it, especially if it adds complexity to the tooling.

**Comment by mpereira** [comment]
For example, if TC changes are required, different operators will have to have different TC build definitions (which different build steps for running the `images/build.sh` script).

**Comment by zmalik** [comment]
we can have `images/build.sh` to build everything and `images/build.sh push` to build and push everything

**Comment by mpereira** [comment]
Sweet!

File: images/cruise-control/scripts/start.sh [hunk 1]
```diff
@@ -0,0 +1,4 @@
+set -eu
+
+echo "${CRUISE_CONTROL_NAMESPACE},${CRUISE_CONTROL_INSTANCE_NAME},/kafkacruisecontrol/" > /opt/cruise-control/cruise-control-ui/dist/static/config.csv
+/opt/cruise-control/kafka-cruise-control-start.sh /etc/cruise-control/config/cruise.properties
```
Comment (porridge) [inline images/cruise-control/scripts/start.sh]:
```suggestion
exec /opt/cruise-control/kafka-cruise-control-start.sh /etc/cruise-control/config/cruise.properties
```

File: images/kafka/Dockerfile [hunk 2]
```diff
@@ -14,6 +14,11 @@ RUN wget https://downloads.mesosphere.com/java/openjdk-jre-8u212b03-hotspot-linu
     mkdir /srv/jdk && \
     tar -xvvf /srv/openjdk-jre-8u212b03-hotspot-linux-x64.tar.gz --strip-components=1 -C /srv/jdk
 
+FROM openjdk:8-jdk as cruise-control
+RUN git clone --branch 2.0.77 https://github.com/linkedin/cruise-control.git && cd cruise-control/
```
Comment (porridge) [inline images/kafka/Dockerfile]: What is the purpose of the `cd`? `git` should exit non-zero if there is trouble, and I'm afraid the `cd` might confuse someone into believing *it* causes the next `RUN` to run in the subdir.

File: images/build.sh [hunk 3]
```diff
@@ -2,9 +2,60 @@
...
 set -eu
 
-source docker-version.sh
-docker image build --build-arg KAFKA_VERSION=${KAFKA_BASE_TECH_VERSION} -t mesosphere/kafka:${DOCKER_IMAGE_VERSION}-${KAFKA_BASE_TECH_VERSION} .
+function print_help() {
+  echo "Usage: ${0} [OPTIONS]"
+  echo
+  echo "OPTIONS:"
+  echo "--push             push the recently build image"
+  echo "--image            docker image to build. valid options are '-image=kafka' and '-image=cruise-control'"
```
Comment (porridge) [inline images/build.sh]:
```suggestion
  echo "--image=IMAGE      docker image to build. valid options are '--image=kafka' and '--image=cruise-control'"
```

File: images/build.sh [hunk 4]
```diff
@@ -2,9 +2,60 @@
...
+        shift
+        ;;
+        *)
+        print_help
+        exit 1
+        ;;
+    esac
+done
+
+if [[ -z "${IMAGE_NAME:-}" ]]; then
```
Comment (porridge) [inline images/build.sh]: Nit: fold this and next `if` into `case`.

File: images/build.sh [hunk 5]
```diff
@@ -2,9 +2,60 @@
...
+  print_help
+fi
+
+if [[ "${IMAGE_NAME}" == "kafka" ]]; then
+  source kafka/docker-version.sh
+  docker image build --build-arg KAFKA_VERSION=${KAFKA_BASE_TECH_VERSION} -t mesosphere/kafka:${KAFKA_DOCKER_IMAGE_VERSION}-${KAFKA_BASE_TECH_VERSION} ./kafka
+elif [[ "${IMAGE_NAME}" == "cruise-control" ]]; then
+  source cruise-control/docker-version.sh
+  docker image build --build-arg CRUISE_CONTROL_VERSION=${CRUISE_CONTROL_VERSION} --build-arg CRUISE_CONTROL_UI_VERSION=${CRUISE_CONTROL_UI_VERSION} \
+    -t mesosphere/cruise-control:${CRUISE_CONTROL_DOCKER_IMAGE_VERSION} ./cruise-control
```
Comment (porridge) [inline images/build.sh]: Why the different versioning scheme for cruise-control image (no base tech version)?

File: images/cruise-control/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,35 @@
+FROM openjdk:8-jdk as build-env
+ARG CRUISE_CONTROL_VERSION
+ARG CRUISE_CONTROL_UI_VERSION
+RUN git clone --branch ${CRUISE_CONTROL_VERSION} https://github.com/linkedin/cruise-control.git && cd cruise-control/
```
Comment (porridge) [inline images/cruise-control/Dockerfile]:
```suggestion
RUN git clone --branch ${CRUISE_CONTROL_VERSION} https://github.com/linkedin/cruise-control.git
```

File: images/cruise-control/Dockerfile [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+WORKDIR cruise-control
+RUN ./gradlew jar --warning-mode=all
+RUN wget https://github.com/linkedin/cruise-control-ui/releases/download/v${CRUISE_CONTROL_UI_VERSION}/cruise-control-ui-${CRUISE_CONTROL_UI_VERSION}.tar.gz && mkdir /cruise-control-ui && \
```
Comment (porridge) [inline images/cruise-control/Dockerfile]:
```suggestion
RUN wget https://github.com/linkedin/cruise-control-ui/releases/download/v${CRUISE_CONTROL_UI_VERSION}/cruise-control-ui-${CRUISE_CONTROL_UI_VERSION}.tar.gz && \
    mkdir /cruise-control-ui && \
```

File: images/cruise-control/Dockerfile [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+    tar -xvvf cruise-control-ui-${CRUISE_CONTROL_UI_VERSION}.tar.gz --strip-components=1 -C /cruise-control-ui && rm cruise-control-ui-${CRUISE_CONTROL_UI_VERSION}.tar.gz
```
Comment (porridge) [inline images/cruise-control/Dockerfile]:
```suggestion
    tar -xvvf cruise-control-ui-${CRUISE_CONTROL_UI_VERSION}.tar.gz --strip-components=1 -C /cruise-control-ui && \
    rm cruise-control-ui-${CRUISE_CONTROL_UI_VERSION}.tar.gz
```

File: images/cruise-control/Dockerfile [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+RUN wget https://downloads.mesosphere.com/java/openjdk-jre-8u212b03-hotspot-linux-x64.tar.gz -P /srv && \
+    mkdir /srv/jdk && \
+    tar -xvvf /srv/openjdk-jre-8u212b03-hotspot-linux-x64.tar.gz --strip-components=1 -C /srv/jdk
```
Comment (porridge) [inline images/cruise-control/Dockerfile]:
```suggestion
    tar -xvvf /srv/openjdk-jre-8u212b03-hotspot-linux-x64.tar.gz --strip-components=1 -C /srv/jdk && \
    rm /srv/openjdk-jre-8u212b03-hotspot-linux-x64.tar.gz
```

File: images/cruise-control/Dockerfile [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+
+# using debian:stretch-20190610-slim
+FROM debian@sha256:9490c476443a3869e39c2897fa66c91daf5dcbbfca53c976dac7bbdc45775b28
+
+RUN apt-get update && apt-get install -y curl openssl netcat procps
```
Comment (porridge) [inline images/cruise-control/Dockerfile]:
```suggestion
RUN apt-get update && apt-get install -y curl openssl netcat procps && \
```

File: images/cruise-control/Dockerfile [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+RUN rm -rf /var/lib/apt/lists/*
```
Comment (porridge) [inline images/cruise-control/Dockerfile]:
```suggestion
    rm -rf /var/lib/apt/lists/*
```
This needs to be in the same layer as the `apt-get update` to be beneficial.

File: images/build.sh (hunk 5 above)
Comment (zmalik) [reply images/build.sh]: yes, this in application and we shouldn't add complexity to this app, unless we are providing an operator for cruise-control itself. 
I would have used something existing and official docker image but as there is no official docker image for cruise-control, went with this one

File: images/kafka/Dockerfile (hunk 2 above)
Comment (zmalik) [reply images/kafka/Dockerfile]: good catch, removing this

File: images/build.sh (hunk 5 above)
Comment (mpereira) [reply images/build.sh]: KUDO C* [does something similar](https://github.com/mesosphere/kudo-cassandra-operator/blob/4517807137461f3308d82cf68bdd619ec9e50bc6/metadata.sh#L48-L62).

The Cassandra image tag ends up being something like `3.11.5-0.1.1` (`3.11.5` is the Cassandra version) and the Prometheus exporter image tag ends up being something like `2.2.1-0.1.1` (`2.2.1` is the Prometheus exporter version).

Comment (zmalik) [reply images/build.sh]: let's use the same pattern here also then: https://github.com/mesosphere/kudo-kafka-operator/pull/29/commits/092e3b824e69b93e8ed04b781805f07c7a208cc2

**Pull request merged**